lib/awful/hotkeys_popup/widget.lua

```diff
@@ -285,6 +295,53 @@ function widget.new(args)
     end
 
 
+    function widget_instance:_abbreviate_awful_keys()
+        -- This method is intended to abbreviate the keys of a merged entry (not
+        -- the modifiers) if and only if the entry consists of five or more
+        -- correlative keys from the same keygroup.
+        --
+        -- For simplicity, it checks only the first keygroup which contains the
+        -- first key. If any of the keys in the merged entry is not in this
+        -- keygroup, or there are any gaps between the keys (e.g. the entry
+        -- contains the 2nd, 3rd, 5th, 6th, and 7th key in
+        -- awful.key.keygroups.numrow, but not the 4th) this method does not try
+        -- to abbreviate the entry.
+        --
+        -- Cheatsheets for external programs are abbreviated by hand where
+        -- applicable: they do not need this method.
+        for _, keys in pairs(self._cached_awful_keys) do
+            for _, params in pairs(keys) do
+                if #params.keylist > 4 then
+                    -- assuming here keygroups will never overlap;
+                    -- if they ever do, another for loop will be necessary:
+                    local keygroup = gtable.find_first_key(self._keygroups, function(_, v)
+                        return not not gtable.hasitem(v, params.keylist[1])
+                    end)
+                    local first, last, count, tally = nil, nil, 0, {}
```

Comment:
i'm not sure, mb you could use `#tally` instead of `count`?

Reply:
It's not reliable. [The manual for 5.3 is verbose but pretty clear:](https://www.lua.org/manual/5.3/manual.html#3.4.7), `#table` may return any key `n` as long as:
* `n` ≥ 0 and integer,
* `table[n]` exists,
* and `table[n+1]` is nil.

If `tally[1]` doesn't exist (if, for example, `2/3/4/5/6` is checked against `numrow`) or there are gaps in the table, `#tally` has undefined behaviour. ([Lua 5.1](https://www.lua.org/manual/5.1/manual.html#2.5.5) explicitly says that if `tally[1]` is nil, `#tally` **may** be 0)

Reply:
yeah, it's indeed not good:

```lua
> t = {}
> #t
0
> t[2] = 123
> #t
0
> t[1] = 123
> #t
2
```